File: Assets/Script/Packet/ChatBroadcastPacket.cs
using System;
using NetworkUtil;

public class ChatBroadcastPacket : ChatPacket
{
    public string Message { get; private set; }

    public override EPacketType PacketType => EPacketType.CHAT_BROADCAST;

    public ChatBroadcastPacket() : base()
    {
        Message = string.Empty;
    }

    public ChatBroadcastPacket(uint uid, ulong timestamp, string message)
        : base(uid, timestamp)
    {
        Message = message;
    }

    public override void Serialize(BinaryWriter writer)
    {
        writer.WriteUInt32(UID);
        writer.WriteUInt64(Timestamp);
        writer.WriteString(Message);

    }

    public override void Deserialize(BinaryReader reader)
    {
        UID = reader.ReadUInt32();
        Timestamp = reader.ReadUInt64();
        Message = reader.ReadString();
    }
}

File: Server_Code/Packet/ChatBroadcastPacket.cpp
#include "ChatBroadcastPacket.h"

ChatBroadcastPacket::ChatBroadcastPacket()
    :ChatBroadcastPacket(0, 0, "")
{
}

ChatBroadcastPacket::ChatBroadcastPacket(uint32_t uid, uint64_t timestamp, const std::string& msg)
    : ChatPacket(EPacketType::CHAT_BROADCAST, uid, timestamp), message(msg)
{
    header.size = sizeof(PacketHeader);
}

std::vector<char> ChatBroadcastPacket::Serialize() const
{
    BinaryWriter writer;
    writer.WriteHeader(header);
    writer.WriteUInt32(uid);
    writer.WriteUInt64(timestamp);
    writer.WriteString(message);

    std::vector<char> buffer = writer.GetBuffer();
    *reinterpret_cast<uint16_t*>(&buffer[2]) = static_cast<uint16_t>(buffer.size());
    return buffer;
}

void ChatBroadcastPacket::Deserialize(const char* buffer, size_t bufferSize)
{
    BinaryReader reader(buffer, bufferSize);
    header = reader.ReadHeader();
    uid = reader.ReadUInt32();
    timestamp = reader.ReadUInt64();
    message = reader.ReadString();
}

bool ChatBroadcastPacket::IsValid(const char* buffer) const
{
    try {
        BinaryReader reader(buffer, sizeof(PacketHeader));
        PacketHeader h = reader.ReadHeader();
        return (static_cast<EPacketType>(h.type) == EPacketType::CHAT_BROADCAST);
    }
    catch (...) {
        return false;
    }
}